backend/test_observer/controllers/artefacts/logic.py

```diff
@@ -0,0 +1,119 @@
+        .order_by(ArtefactBuild.architecture, ArtefactBuild.revision.desc())
+        .options(
+            joinedload(ArtefactBuild.test_executions).joinedload(
+                TestExecution.environment
+            )
+        )
+        .all()
+    )
+
+    return orm_builds
+
+
+def get_statuses_ids(builds: list[ArtefactBuild]) -> Iterator[int]:
+    for build in builds:
+        for test_execution in build.test_executions:
+            c3_link = test_execution.c3_link
+            if c3_link and (status_id := _parse_status_id_from_c3_link(c3_link)):
+                yield status_id
+
+
+def get_reports_ids(submissions_statuses: Iterable[SubmissionStatus]) -> Iterator[int]:
+    return (ss.report_id for ss in submissions_statuses if ss.report_id is not None)
+
+
+def construct_dto_builds(
+    builds: list[ArtefactBuild],
+    submissions_statuses: dict[int, SubmissionStatus],
+    reports: dict[int, Report],
+) -> Iterator[ArtefactBuildDTO]:
+    for build in builds:
+        test_executions_dto = []
+        for test_execution in build.test_executions:
+            test_executions_dto.append(
+                TestExecutionDTO(
+                    id=test_execution.id,
+                    jenkins_link=test_execution.jenkins_link,
+                    c3_link=test_execution.c3_link,
+                    environment=EnvironmentDTO.model_validate(
+                        test_execution.environment
+                    ),
+                    status=_derive_test_execution_status(
+                        test_execution.c3_link, submissions_statuses, reports
+                    ),
+                )
+            )
+
+        build_dto = ArtefactBuildDTO(
+            id=build.id,
+            architecture=build.architecture,
+            revision=build.revision,
+            test_executions=test_executions_dto,
+        )
+
+        yield build_dto
+
+
+def _derive_test_execution_status(
+    c3_link: str | None,
+    submissions_statuses: dict[int, SubmissionStatus],
+    reports: dict[int, Report],
+) -> TestExecutionStatus:
+    if c3_link is None:
+        return TestExecutionStatusLogic.NO_C3_LINK
+    if (status_id := _parse_status_id_from_c3_link(c3_link)) is None:
+        return TestExecutionStatusLogic.NO_STATUS_ID_IN_C3_LINK
+    if (submission_status := submissions_statuses.get(status_id)) is None:
+        return TestExecutionStatusLogic.SUBMISSION_STATUS_NOT_FOUND
+    if submission_status.status == SubmissionProcessingStatus.FAIL:
+        return TestExecutionStatusLogic.SUBMISSION_STATUS_IS_FAIL
+    if (report_id := submission_status.report_id) is None:
+        return TestExecutionStatusLogic.NO_REPORT_ID_IN_SUBMISSION_STATUS
+    if (report := reports.get(report_id)) is None:
+        return TestExecutionStatusLogic.REPORT_NOT_FOUND
+    if report.failed_test_count == 0:
+        return TestExecutionStatusLogic.NO_FAILED_TESTS
+    else:
+        return TestExecutionStatusLogic.SOME_FAILED_TESTS
```

Comment:
you can remove else after return statement. however, if you want to keep else in this if-else sequence, I guess it's logical to use `elif` instead of just `if` stating from the 2nd if